Feature/hatch flower

kc5tjn opened the pull request:
The HatchFlower Module is ready for testing.

kc5tjn committed:
Implement HatchFlower Module.

kc5tjn committed:
Change order of Capture to push solenoid transitions.

kc5tjn committed:
Added the HatchFlower to Robot teleopInit

flybotix commented on robot/src/main/java/us/ilite/robot/modules/HatchFlower.java:
```diff
@@ -1,73 +1,353 @@
+    // private final Codex<Double, ELogitech310> mController;
+
+    private CommandQueue mCurrentCommandQueue;
+
```
We'll always start the match with either a hatch or cargo.  So the initial state for the hatch will always be 'CAPTURE'.

flybotix commented on robot/src/main/java/us/ilite/robot/modules/HatchFlower.java:
```diff
@@ -1,73 +1,353 @@
+
+        @Override
+        public boolean update(double pNow) {
+            // always return done, delays handled by the Delay command
```
The 'grab' command is never 'finished'.  It is only 'cancelled'.  This is probably why we'd here a click as the solenoid opened during command init, but didn't stay open afterwards.  Recommend moving the solenoid command to the update() method, and always returning false.

kc5tjn committed:
Changed the initial state to CAPTURE per JKnight.  Changed logic to allow re-issue of the capture command, if already in the capture state. Added some clarifying comments.

kc5tjn commented:
Jesse,

I made the change to the initial state to CAPTURE for the HatchFlower.  As
for the comment "The 'grab' command is never 'finished'. It is only
'cancelled'......"  I was assuming that the solenoids would hold the last
commanded state, is this not the case?

If the solenoids hold the commanded state, then the idea was that the
commands would set the state on init, then complete on the first update.
The CommandQueue should ensure that the steps execute sequentially.

Thus, when a capture is commanded, the HatchFlower would set the solenoid
states so that the hatch is captured (grabbed), and leave the HatchFlower
in the CAPTURE state, which will persist until the PUSH is commanded.
Since the Push is a transient action, the HatchFlower will only be in the
PUSH state while the push is executing.  The push sequence of commands will
leave the HatchFlower in the RELEASE state, with both solenoids in the
release/reset state.

If I always return false for the grab command then no other commands would
be executed.

Am I missing something?  If you want to talk about it you can reach me at:
jmshapiro@iname.com
571-215-5418

Thanks,
Jeff

On Mon, Jan 28, 2019 at 10:34 AM Jesse Knight <notifications@github.com>
wrote:

> *@flybotix* requested changes on this pull request.
> ------------------------------
>
> In robot/src/main/java/us/ilite/robot/modules/HatchFlower.java
> <https://github.com/iliterobotics/FRC-Robot-2019/pull/34#discussion_r251463659>
> :
>
> >      private ILog mLog = Logger.createLog(HatchFlower.class);
>
>      private Solenoid solenoidOpenClose;
>      private Solenoid solenoidExtension;
>
> -    public enum SoleniodStates {
> -        // TODO hold solenoid states for each state
> -        CAPTURE,
> -        RELEASE,
> -        PUSH;
> +    // Needed for prototype testing
> +    // private final Codex<Double, ELogitech310> mController;
> +
> +    private CommandQueue mCurrentCommandQueue;
> +
>
> We'll always start the match with either a hatch or cargo. So the initial
> state for the hatch will always be 'CAPTURE'.
> ------------------------------
>
> In robot/src/main/java/us/ilite/robot/modules/HatchFlower.java
> <https://github.com/iliterobotics/FRC-Robot-2019/pull/34#discussion_r251464841>
> :
>
> > +    public class GrabSolenoidCommand implements ICommand {
> +
> +        GrabberState state;
> +
> +        GrabSolenoidCommand(GrabberState state) {
> +            this.state = state;
> +        }
> +
> +        @Override
> +        public void init(double pNow) {
> +            solenoidOpenClose.set(this.state.grabber);
> +        }
> +
> +        @Override
> +        public boolean update(double pNow) {
> +            // always return done, delays handled by the Delay command
>
> The 'grab' command is never 'finished'. It is only 'cancelled'. This is
> probably why we'd here a click as the solenoid opened during command init,
> but didn't stay open afterwards. Recommend moving the solenoid command to
> the update() method, and always returning false.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/iliterobotics/FRC-Robot-2019/pull/34#pullrequestreview-197097904>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AqOcr8-Qwz8GpJS5awA6_4F8eB7_eFU-ks5vHxibgaJpZM4aU50w>
> .
>

oconnelc commented:
@kc5tjn I'm not sure I understand your comment about the Grab Command. The Grab Command is added to the: 
`ParallelCommand step2= new ParallelCommand(step2List);`

All of the commands in this parallel command will get executed on each update. If the method returns true, then the command is done and will get removed from the queue. Since we always returning false, the command will never be returned but the other commands will be executed.

ChangJoseph committed:
Added controls to HatchFlower

Need HatchFlower param when calling DriverInput constructor
Temporarily changed solenoid IDs in HatchFlower
Added mHatchFlower in DriverInputTest because of new constructor

StephenWelch commented on robot/src/main/java/us/ilite/robot/modules/HatchFlower.java:
```diff
@@ -1,73 +1,366 @@
+        // TODO Do we need to pass the CAN Addresses in via the constructor?
+        // grabSolenoid = new Solenoid(SystemSettings.kHatchFlowerOpenCloseSolenoidAddress);
+        // pushSolenoid = new Solenoid(SystemSettings.kHatchFlowerExtensionSolenoidAddress);
+        grabSolenoid = new Solenoid(4); //Ball Grabber/Holder (2016 robot)
```
Please re-comment these lines before pushing next time

StephenWelch committed:
Minor syntax and variable name changes

StephenWelch committed:
Merge remote-tracking branch 'origin/master' into feature/HatchFlower

ChangJoseph committed:
Constant name & added mHatchFlower to mRunningModules
